source/opt/folding_rules.cpp

```diff
@@ -113,6 +207,49 @@ FoldingRule InsertFeedingExtract() {
     return true;
   };
 }
+
+FoldingRule RedundantPhi() {
+  // An OpPhi instruction where all values are the same or the result of the phi
+  // itself, can be replaced by the value itself.
+  return
+      [](ir::Instruction* inst, const std::vector<const analysis::Constant*>&) {
+        assert(inst->opcode() == SpvOpPhi && "Wrong opcode.  Should be OpPhi.");
+
+        ir::IRContext* context = inst->context();
+        analysis::DefUseManager* def_use_mgr = context->get_def_use_mgr();
+
+        uint32_t incoming_value = 0;
+
+        for (uint32_t i = 0; i < inst->NumInOperands(); i += 2) {
+          uint32_t op_id = inst->GetSingleWordInOperand(i);
+          if (op_id == inst->result_id()) {
+            continue;
+          }
+
+          ir::Instruction* op_inst = def_use_mgr->GetDef(op_id);
+          if (op_inst->opcode() == SpvOpUndef) {
+            continue;
```

Comment:
This case is probably wrong.  We might need to figure out something more precise because I do see Undef's in some real cases.